frontend-react/src/AppRouter.tsx

```diff
@@ -41,6 +43,50 @@ export enum FeatureName {
 }
 
 export const AppRouter = () => {
+    const {
+        sessionStartTime,
+        setSessionStartTime,
+        sessionTimeAggregate,
+        setSessionTimeAggregate,
+    } = useSessionContext();
+
+    useEffect(() => {
+        const onUnload = () => {
+            trackAppInsightEvent("session", {
+                sessionLength: sessionTimeAggregate,
+            });
+        };
+
+        const onVisibilityChange = () => {
+            if (document.visibilityState === "hidden") {
+                setSessionTimeAggregate(
+                    // Typescript doesn't like subtracting Date objects from
+                    // each other as there's implicit type coercion, so
+                    // we explicitly turn them into integers for subtraction
+                    // and return a value in seconds.
+                    (new Date().getTime() - sessionStartTime.getTime()) / 1000 +
+                        sessionTimeAggregate
+                );
+            } else if (document.visibilityState === "visible") {
+                setSessionStartTime(new Date());
+            }
+        };
+
+        window.addEventListener("beforeunload", onUnload);
+        window.addEventListener("visibilitychange", onVisibilityChange);
+        return () => {
+            // We MUST remove the event listeners as the dependency values
+            // update or else we'll have 100s of event listeners on the page
+            // all with stale values
+            window.removeEventListener("beforeunload", onUnload);
+            window.removeEventListener("visibilitychange", onVisibilityChange);
+        };
+    }, [
+        sessionStartTime,
```

Comment:
Why does this effect need to rerun whenever these change?  I think we could just make it a one-time* run to bind these effects on mount with `[]`, no?

* Not exactly a one-time run with React v18 strict mode, but it should be good enough for us since we're not there yet 😆

Reply:
Bumping this question 🐢

Reply:
Test comment!

Reply:
I think the issue here is that we're expecting a state change to flush out a re-render, but nothing visual is actually updating hence the stale values.  For this purpose, I'd recommend using `ref`s instead:

```ts
export const AppRouter = () => {
    // I think it's easier to treat these as `getTime()` return values (milliseconds)
    // so 1) there's a consistent unit and 2) we can perform basic math operations on them
    // and then convert to seconds (x / 1000) when we actually need to submit to AI
    const sessionStartTime = useRef<number>(new Date().getTime());
    const sessionTimeAggregate = useRef<number>(0);

    useEffect(() => {
        const onUnload = () => {
            trackAppInsightEvent(EventName.SESSION, {
                sessionLength:
                    (new Date().getTime() -
                        sessionStartTime.current +
                        sessionTimeAggregate.current) /
                    1000,
            });
        };

        const onVisibilityChange = () => {
            if (document.visibilityState === "hidden") {
                sessionTimeAggregate.current =
                    new Date().getTime() -
                    sessionStartTime.current +
                    sessionTimeAggregate.current;
            } else if (document.visibilityState === "visible") {
                sessionStartTime.current = new Date().getTime();
            }
        };

        // bind + unbind stuff here
    }, []);

    //...otherStuff
};
```

This is untested, but I think that should work since refs exist outside of the state reconciler!